engine/src/test/java/org/camunda/bpm/engine/test/bpmn/tasklistener/TaskListenerEventLifecycleTest.java

```diff
@@ -0,0 +1,586 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.engine.test.bpmn.tasklistener;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+import static org.hamcrest.CoreMatchers.is;
+
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.camunda.bpm.engine.delegate.DelegateTask;
+import org.camunda.bpm.engine.delegate.TaskListener;
+import org.camunda.bpm.engine.history.HistoricVariableInstanceQuery;
+import org.camunda.bpm.engine.runtime.JobQuery;
+import org.camunda.bpm.engine.runtime.ProcessInstance;
+import org.camunda.bpm.engine.task.Task;
+import org.camunda.bpm.engine.task.TaskQuery;
+import org.camunda.bpm.engine.test.bpmn.tasklistener.util.AssigneeAssignment;
+import org.camunda.bpm.engine.test.bpmn.tasklistener.util.CandidateUserAssignment;
+import org.camunda.bpm.engine.test.bpmn.tasklistener.util.CompletingTaskListener;
+import org.camunda.bpm.engine.test.bpmn.tasklistener.util.RecorderTaskListener;
+import org.camunda.bpm.model.bpmn.Bpmn;
+import org.camunda.bpm.model.bpmn.BpmnModelInstance;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TaskListenerEventLifecycleTest extends AbstractTaskListenerTest{
+  /*
+  Testing Task Event chains to validate event lifecycle order
+
+  Note: completing tasks inside TaskListeners breaks the global task event lifecycle. However,
+  task events are still fired in the right order inside the listener "scope".
+   */
+
+  protected static final String[] TRACKED_EVENTS = {
+      TaskListener.EVENTNAME_CREATE,
+      TaskListener.EVENTNAME_UPDATE,
+      TaskListener.EVENTNAME_ASSIGNMENT,
+      TaskListener.EVENTNAME_COMPLETE,
+      TaskListener.EVENTNAME_DELETE
+  };
+
+  // CREATE phase
+
+  @Test
+  public void shouldFireCreateEventBeforeAssignmentEventWhenTaskCreated() {
+    // given
+    createAndDeployModelWithTaskEventsRecorderOnUserTaskWithAssignee("kermit", TRACKED_EVENTS);
+
+    // when
+    runtimeService.startProcessInstanceByKey("process");
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+    assertThat(orderedEvents.size()).isEqualTo(2);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT);
+  }
+
+  @Test
+  public void shouldNotFireUpdateEventWhenTaskIsCreated() {
+    // given
+    createAndDeployModelWithTaskEventsRecorderOnUserTaskWithAssignee("kermit", TRACKED_EVENTS);
```

Comment:
Shouldn't there be a create listener that updates properties here?
Otherwise, the test is identical to the one above.

Reply:
Since the model has an assignee already set, this would fire an update event, if the task was already created. But you're right, the tests are completely the same. I can just leave one, and rename it to `shouldOnlyFireCreateAndAssignmentEventsWhenTaskIsCreated`. What do you think?

Reply:
Yeah, that sounds about right as well. 👍